sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/SetLocaleForDataTypes.cs

```diff
@@ -0,0 +1,231 @@
+
+                    compilationStartContext.RegisterSyntaxNodeActionInNonGenerated(
+                        c =>
+                        {
+                            var objectCreation = (ObjectCreationExpressionSyntax)c.Node;
+                            var objectType = c.SemanticModel.GetSymbolInfo(objectCreation.Type).Symbol as ITypeSymbol;
+
+                            if (objectType == null ||
+                                !objectType.IsAny(invalidTypes))
+                            {
+                                return;
+                            }
+
+                            var variableSyntax = GetAssignedToSyntax(objectCreation);
+                            if (variableSyntax == null)
+                            {
+                                return;
+                            }
+
+                            var variableSymbol = variableSyntax is IdentifierNameSyntax
+                                ? c.SemanticModel.GetSymbolInfo(variableSyntax).Symbol
+                                : c.SemanticModel.GetDeclaredSymbol(variableSyntax);
+                            if (variableSymbol != null)
+                            {
+                                symbolsWhereTypeIsCreated.Add(variableSymbol.ToSymbolWithSyntax(objectCreation));
+                            }
+                        }, SyntaxKind.ObjectCreationExpression);
+
+                    compilationStartContext.RegisterSyntaxNodeActionInNonGenerated(
+                        c =>
+                        {
+                            var assignmentExpression = (AssignmentExpressionSyntax)c.Node;
+                            var propertySymbol = GetPropertySymbol(assignmentExpression, c.SemanticModel);
+
+                            if (propertySymbol == null ||
+                                !propertySymbol.ContainingType.IsAny(invalidTypes) ||
+                                propertySymbol.Name != "Locale")
+                            {
+                                return;
+                            }
+
+                            var variableSymbol = GetAccessedVariable(assignmentExpression, c.SemanticModel);
+                            if (variableSymbol != null)
+                            {
+                                symbolsWhereLocaleIsSet.Add(variableSymbol);
+                            }
+                        }, SyntaxKind.SimpleAssignmentExpression);
+
+                    compilationStartContext.RegisterCompilationEndAction(
+                        c =>
+                        {
+                            var invalidDataTypeCreation = symbolsWhereTypeIsCreated
+                                .Where(x => !symbolsWhereLocaleIsSet.Contains(x.Symbol));
+
+                            foreach (var invalidCreation in invalidDataTypeCreation)
+                            {
+                                var typeName = GetConstructedTypeName(invalidCreation.Symbol);
```

Comment:
You could use `GetSymbolType(invalidCreation.Symbol)?.Name` instead (and remove your method)